featurebyte/session/session_helper.py

```diff
@@ -180,22 +183,40 @@ async def execute_feature_query_set(
     Optional[pd.DataFrame]
     """
     session = session_handler.session
-    total_num_queries = len(feature_query_set.feature_queries) + 1
+    source_info = session.get_source_info()
+    total_num_nodes = feature_query_set.get_num_nodes()
     materialized_feature_table = []
     processed = 0
 
-    async def _progress_callback() -> None:
+    # Allocate 90% of the progress to feature queries and 10% to the final output query
+    if progress_callback is None:
+        feature_queries_progress_callback, output_query_progress_callback = None, None
+    else:
+        feature_queries_progress_callback, output_query_progress_callback = (
+            divide_progress_callback(progress_callback, 90)
+        )
+
+    async def _progress_callback(num_nodes_completed: int) -> None:
         nonlocal processed
-        processed += 1
-        if progress_callback:
-            await progress_callback(
-                int(100 * processed / total_num_queries),
+        processed += num_nodes_completed
+        if feature_queries_progress_callback:
+            await feature_queries_progress_callback(
+                int(100 * processed / total_num_nodes),
                 feature_query_set.progress_message,
             )
 
+    generator = feature_query_set.feature_query_generator
```

Comment:
The main change is that the feature queries to be executed will be generated here within `execute_feature_query_set`. This is to allow this function to retry failed queries with smaller batch sizes later.